### ISSUE TITLE
Foreign key arrows not linking to the correct row

### ISSUE THREAD

**Issue by Firehed** [sequelpro/sequelpro#2792]
When clicking the arrow on a foreign keyed column to view the linked row, the correct row is not displayed. It appears that SP is not filling in the correct filter information once the related table is selected.

Happening on build 5190 (f741be7), connecting to a MySQL 5.7 server over SSH. Please let me know if you need additional information!

**Comment by dmoagx** [comment]
Yeah, I can reproduce that.
From my initial debugging the culprit here is the history controller, which overwrites the restore info after it has been configured correctly.
There may also be multiple race conditions involved here.

**Comment by dmoagx** [comment]
Looking deeper, the issue is with `-[SPTablesList selectItemWithName:]`: If this method is called from a bg thread (as it was in 1.1.2) it will result in a call chain that synchronously blocks the thread until after the table switch. However, when called from the main thread (as it is now) it will act like an asynchronous method and return right away. That way the "linked data load" thread runs its cleanup code before the filters have actually been applied.

Now, this puts us in a bind, as the 1.1.2 way of calling a UI method (`selectItemWithName:`) on a background thread was also severely broken, but we don't have the code infrastructure for handling the current async code, either.

**Comment by dmoagx** [comment]
For illustration purposes, this is how it worked in 1.1.2 (x axis is time):
![112](https://cloud.githubusercontent.com/assets/5793864/26528984/a4bdd6d8-43b7-11e7-8e8a-d3068fe6b547.png)

This is how it works ATM:
![current](https://cloud.githubusercontent.com/assets/5793864/26528986/ad73fd34-43b7-11e7-9459-85df13d752d9.png)

The `updateHistoryEntries:` call I have marked either needs to be the last call (as in 1.1.2) or we have to think about an entirely different approach.

**Comment by dmoagx** [comment]
While looking for fixes I realized that the problem actually affects the whole history controller, because back/forward navigation is also based on the assumption that the whole call tree would be synchronous (which it isn't). 😩

**Comment by jmisur** [comment]
Hi, this functionality is broken for quite some time in nightly builds. It's one of the main great usability features Sequel Pro offers. If someone get's used to it it's painful to see it broken. Could you please consider fixing it :) Great job with the project otherwise 👍

**Comment by enekochan** [comment]
Still not working in df10e20 This makes Sequel Pro almost unusable for me and probably for many others :(

**Comment by enekochan** [comment]
Downloading a lot of [test builds](https://sequelpro.com/test-builds) (and doing some divide and conquer) I think I've found that the bug starts appearing between the commits dfab0cd76c and bed1cff292.

**Comment by Kaspik** [comment]
@enekochan Can you please verify it's still broken on latest nightly build from commit `37f62834a6`? Thank you!

**Comment by cmorbitzer** [comment]
I can verify that it works if the table does not change (foreign key to same table) but it still does not work if the table is different.

**Comment by enekochan** [comment]
@Kaspik Yes it happens in 37f62834a6 and also in current latest build c88e7c5b74.

**Comment by dmoagx** [comment]
Well, of course it does. ;)
I explained the cause in detail above and so far nobody has fixed that.

**Comment by sauron918** [comment]
**Still happening** on build 5446 (97c1b857)
